### ISSUE TITLE
guildMemberUpdate new member and old member are the same object

### ISSUE THREAD

**Issue by Linker-123** [discordjs/discord.js#4800]
I was making audit logs system for my bot and encountered a problem the **oldMember.user** object is the same is **newMember.user** object.
i tried doing `console.log(oldMember.user.username + '|' + newMember.user.username)` and the results both were the new username of the user!

**Comment by NotSugden** [comment]
i don't think `guildMemberUpdate` fires for `username` changes, are you perhaps looking for [`GuildMember#nickname`](https://discord.js.org/#/docs/main/stable/class/GuildMember?scrollTo=nickname)?

[`GuildMember#User`](https://discord.js.org/#/docs/main/stable/class/GuildMember?scrollTo=user) is a `User` object, which is different from a `GuildMember` object, it represents the user as a whole, not just for one specific guild

**Comment by Crazyshooting** [comment]
> ~i don't think `guildMemberUpdate` fires for `username` changes~ it does fire, but still, are you perhaps looking for [GuildMember#nickname](https://discord.js.org/#/docs/main/stable/class/GuildMember?scrollTo=nickname)?
> 
> [GuildMember#User](https://discord.js.org/#/docs/main/stable/class/GuildMember?scrollTo=user) is a `User` object, which is different from a `GuildMember` object, it represents the user as a whole, not just for one specific guild

no i was looking for the username change

**Comment by NotSugden** [comment]
ah, seem to have found the issue
the GUILD_MEMBER_UPDATE handler https://github.com/discordjs/discord.js/blob/dfd63bdb6b73c39bd071689561f42dce3827fc14/src/client/websocket/handlers/GUILD_MEMBER_UPDATE.js#L16) calls `GuildMember#_update` which calls the `_patch` method, and updates the user on the member https://github.com/discordjs/discord.js/blob/dfd63bdb6b73c39bd071689561f42dce3827fc14/src/structures/GuildMember.js#L79

**Comment by almostSouji** [comment]
I have stumbled across this issue before but forgot to make a trackable issue for it, thanks for bringing it up again! I will try to shortly summarize and explain what i found out:

### Upfront conclusion

Discord sends both PRESENCE_UPDATE as well as GUILD_MEMBER_UPDATE per guild the bot shares with the member if the underlying user information is changed.

- On the PU we determine the change and apply it to the cache
- On the GMU the underlying user is already changed from the PU before and we can not emit the old user with the old member

We can not just use the data discord provides either as the GMU will not hold "old" data but just the updated version. The old variants of structures are always retrieved from cache due to that reason.

In this case the cache is already changed and we have no way of getting the old user in the GMU which would be the cached version (old user) from before applyint the PU.

### Considered solutions

A separate cache for the old user `_frozenUser`/`_frozenPresence` could be considered however comes with its own set of challenges, as the timeline for these changes PU -> GMU times n (where n is unknown and even in between updates for n there could technically be another PU which then comes with its own series of GMUs) is not specified or guaranteed.

I have thought a lot about how we could solve this but have not arrived at a feasible solution so far.

### Reference

> A user's presence is their current state on a guild. This event is sent when a user's presence or info, such as name or avatar, is updated.

> The user object within this event can be partial, the only field which must be sent is the id field, everything else is optional. Along with this limitation, no fields are required, and the types of the fields are not validated. Your client should expect any combination of fields and types within this event.
<https://discord.com/developers/docs/topics/gateway#presence-update>

> Sent when a guild member is updated. This will also fire when the user object of a guild member changes.
<https://discord.com/developers/docs/topics/gateway#guild-member-update>

Interestingly enough there is also

> User Update
Sent when properties about the user change. Inner payload is a user object.
<https://discord.com/developers/docs/topics/gateway#user-update>

Which does not emit when changing a username of the test account - however this has emitted when my bot set its username to the current commit (as my test cluster does). Maybe this means USER_UPDATE only emits when the logged in user changes? I am really not too sure about this one.

### Testing

<details>
<summary>Click to show</summary>

I have decided to attach a listener to USER_UPDATE as well, however it does not emit when changing the test accounts username.

I changed my alts username from `maybeSouji` to `alsoSouji`. The bot is present on one server where my alt is present, meaning only one GUILD_MEMBER_UPDATE instance is expected.

```js
const observed = '175179625880616960';

client.ws.on('GUILD_MEMBER_UPDATE', d => {
	if (d.user.id !== observed) return;
	console.log('GUILD_MEMBER_UPDATE', d);
});

client.ws.on('PRESENCE_UPDATE', d => {
	if (d.user.id !== observed) return;
	console.log('PRESENCE_UPDATE', d);
});

client.ws.on('USER_UPDATE', d => {
	if (d.id !== observed) return;
	console.log('USER_UPDATE', d);
});

client.on('guildMemberUpdate', (o, n) => {
	if (o.id !== observed) return;
	console.log('gMU (old)', o.user.username);
	console.log('gMU (new)', n.user.username);
});

client.on('userUpdate', (o, n) => {
	if (o.id !== observed) return;
	console.log('uU (old)', o.username);
	console.log('uU (new)', n.username);
});
```

</details>

<details>
<summary>Logs</summary>

```
PRESENCE_UPDATE {
  user: {
    username: 'alsoSouji',
    public_flags: 256,
    id: '175179625880616960',
    discriminator: '4945',
    avatar: '313acc20bef02fa4e6b33276915ae719'
  },
  status: 'online',
  roles: [ '345342066798231553' ],
  premium_since: null,
  nick: null,
  guild_id: '345341901051920395',
  game: null,
  client_status: { desktop: 'online' },       
  activities: []
}
uU (old) maybeSouji
uU (new) alsoSouji
GUILD_MEMBER_UPDATE {
  user: {
    username: 'alsoSouji',
    public_flags: 256,
    id: '175179625880616960',
    discriminator: '4945',
    avatar: '313acc20bef02fa4e6b33276915ae719'
  },
  roles: [ '345342066798231553' ],
  premium_since: null,
  nick: null,
  mute: false,
  joined_at: '2020-07-18T18:49:59.260938+00:00',
  hoisted_role: null,
  guild_id: '345341901051920395',
  deaf: false
}
gMU (old) alsoSouji
gMU (new) alsoSouji
```

</details>

### Interpretation of the test

1. PRESENCE_UPDATE (ws) arrives with the new username "alsoSouji"
2. userUpdate (client) emits with the name changes "maybeSouji" -> "alsoSouji"
3. GUILD_MEMBERUPDATE (ws) arrives with the new username "alsoSouji"
4. guildMemberUpdate (client) emits with the already updated user in both old and new "alsoSouji", "alsoSouji"

**Comment by advaith1** [comment]
yeah `USER_UPDATE` only fires when the bot changes, discord.js's `userUpdate` event is triggered by user changes in `USER_UPDATE`, `PRESENCE_UPDATE`, and now also `GUILD_MEMBER_UPDATE`, as Discord started triggering GMU for user updates a few months ago.

**Comment by SpaceEEC** [comment]
To avoid emitting this update event if there was no actual update for the member itself, we could implement a `GuildMember#equals` and use that to guard emitting the `guildMemberUpdate` event, like we already do for emojis and users.

**Comment by ckohen** [comment]
I guess I missed this issue when creating #5161, it seems the desired solution is actually the exact opposite of what I had done. Keep emitting userUpdate and simply guard against emitting GMU and PU when no other parameters have changed.
If that is the case, I'd more than happily turn that PR around to implement this solution.

The whole secondary user cache thing was heavily downvoted on 5161 so it seems like a non-solution.

**Comment by ckohen** [comment]
This should now be resolved! #5655
